File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set()
        left = 0
        max_length = 0

        for right in range(len(s)):
            # If char is already in set, move left pointer to remove duplicates
            while s[right] in char_set:
                char_set.remove(s[left])
                left += 1
            
            # Add the current char
            char_set.add(s[right])
            
            # Update max length
            max_length = max(max_length, right - left + 1)

        return max_length
        

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let start = 0
    let windowMap = new Map()
    let maxLength = 0
    for(let end = 0; end < s.length; end++){
        let endChar = s[end]
        if(windowMap.has(endChar)){
            start = Math.max(start,windowMap.get(endChar)+1)
        }
        windowMap.set(endChar,end)
        maxLength = Math.max(maxLength,end-start+1)
    }
    return maxLength
};